File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    queue = deque()
    
    height = len(maps) # 세로
    width = len(maps[0]) # 가로

    # 북 남 동 서 이동 좌표
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
        
    queue.append((0, 0))
    
    while queue:
        x, y = queue.popleft()
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            # 범위 체크
            if nx < 0 or nx >= height or ny < 0 or ny >= width:
                continue
                
            # 벽(0)인지 확인
            if maps[nx][ny] == 0:
                continue
                
            # 방문 가능한 길(1)이라면 거리 계산
            if maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                queue.append((nx, ny))
    
    # 최종 목적지에 도달 못했다면 -1 반환
    return maps[height-1][width-1] if maps[height-1][width-1] != 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int height = maps.length-1;
        int width = maps[0].length-1;

        // 방향: 하, 상, 우, 좌
        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        //Queue<int[]> queue = new LinkedList<>();
        //queue.offer(new int[]{0, 0});
        
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0,0});
        
        
        while (!queue.isEmpty()) {
            int[] now = queue.poll();
            int x = now[0];
            int y = now[1];

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx > height || ny > width || nx < 0 || ny <0){
                    continue;
                }
                
                if (maps[nx][ny] !=1){
                    continue;
                }
                
                maps[nx][ny] = maps[x][y] + 1;
                queue.offer(new int[]{nx,ny});
            }
        }

        // 도착지 값이 여전히 1이면 도달 못한 것
        int answer = maps[height][width];
        return answer == 1 ? -1 : answer;
    }
}
